### PR TITLE
Fully deprecate the symbol display builder, reimplement in terms of node builder

### PR DESCRIPTION
**Pull request by weswigham**
While working on the new declaration emitter, I recognized that there were certain parts of the current declaration emitter which relied on the symbol display builder. For the most part, we had heavily duplicated logic between the symbol display builder and the node builder. The PR fully removes the symbol display builder; replacing it with the node builder and a printer, after upgrading our emitter to do some of its writes in the more granular fashion that the symbol display writer used to expect, and altering the node builder to be capable of the little bit of extra reporting the declaration emitter expects.

This enables me to use the node builder (as one would expect) in the new declaration emitter in a similar fashion to how it is now used in the current declaration emitter.

While I've tried to keep the baseline changes to the absolute minimum, there are a few correctness changes, and a few whitespace changes. The correctness changes are for mapped types - previously when the constraint was written, it would write the declared constraint, rather than the instantiated constraint (which was noticeable in declaration emit baselines, so this has now been fixed). The whitespace changes have to do with the emitter - parameter, type parameter, and type argument lists were marked as both `SingleLine` and `Indented` - causing objects written across multiple lines to have way more indents than they should - these are no longer marked as `Indented`, and so no longer randomly add indentation (even though they should never add newlines on their own). This improved some of our js baselines to actually match the input, which is a good thing (in addition to making the declaration output more closely match old output). Other minor changes are that mapped types are written on a single line when synthesized, rather than on multiple lines, and that the declaration emitter more aggressively uses the `typeof` operator where possible (as before a small bug in the symbol display builder was preventing it from finding some local aliases).

As the `SymbolDisplayBuilder` (and `SymbolWriter`, which has been subsumed into `EmitTextWriter`) was part of our public API, this is a breaking change - however new `XToString` APIs exist which take a writer and can be used to polyfill the functionality.

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -2631,5 +2631,9 @@ namespace ts {
         return (arg: T) => f(arg) && g(arg);
     }
 
+    export function or<T>(f: (arg: T) => boolean, g: (arg: T) => boolean) {
```
Comment (sandersn) [inline src/compiler/core.ts]: `<||>` is much better name for this function bro

File: src/compiler/types.ts [hunk 2]
```diff
@@ -2797,6 +2772,8 @@ namespace ts {
         InArrayType                     = 1 << 15,  // Writing an array element type
         UseAliasDefinedOutsideCurrentScope = 1 << 16, // For a `type T = ... ` defined in a different file, write `T` instead of its value,
                                                       // even though `T` can't be accessed in the current scope.
+        MultilineObjectLiterals         = 1 << 17,   // Always print object literals across multiple lines (only used to map into node builder flags)
```
Comment (sandersn) [inline src/compiler/types.ts]: which flags did the SymbolWriter use? I expected one enum to go away, and instead I see only additions to enums around here.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -623,7 +623,7 @@ namespace ts {
         /*@internal*/ autoGenerateKind?: GeneratedIdentifierKind; // Specifies whether to auto-generate the text for an identifier.
         /*@internal*/ autoGenerateId?: number;                    // Ensures unique generated identifiers get unique names, but clones get the same name.
         isInJSDocNamespace?: boolean;                             // if the node is a member in a JSDoc namespace
-        /*@internal*/ typeArguments?: NodeArray<TypeNode>;        // Only defined on synthesized nodes. Though not syntactically valid, used in emitting diagnostics.
+        /*@internal*/ typeArguments?: NodeArray<TypeNode | TypeParameterDeclaration>; // Only defined on synthesized nodes. Though not syntactically valid, used in emitting diagnostics.
```
Comment (sandersn) [inline src/compiler/types.ts]: why are TypeParameterDeclarations here suddenly? Is this related to the 'syntactically invalid' comment?

Comment (weswigham) [reply src/compiler/types.ts]: ...kinda. The `"syntactically invalid"` comment is referencing us making identifiers with text which is not valid identifier text, like `["foo"]` (and is the root cause of having odd quickinfo like `Foo[["bar"]]` for computed properties sometimes). I don't want to fix that as part of this, though, so just left the comment as a reminder that it was fishy, so it'd be fixed later.

File: src/compiler/types.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/types.ts]: The `SymbolWriter` used `SymbolFormatFlags` and `TypeFormatFlags`. Both are still in use as part of the public API of the `XToString` functions, however if you'd like I can replace them all with `NodeBuilderFlags` directly - it'd make the code a lot simpler, for sure.

File: src/compiler/core.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/core.ts]: If you can write that with emoji instead of reserved characters, it's doable. Like 🤷‍♂️ is kinda like that. Could use that. 😄

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -2302,35 +2394,42 @@ namespace ts {
...
             if (modifiers && modifiers.length) {
                 emitList(node, modifiers, ListFormat.Modifiers);
-                write(" ");
+                writeSpace(" ");
             }
         }
 
-        function emitWithPrefix(prefix: string, node: Node) {
-            emitNodeWithPrefix(prefix, node, emit);
+        function emitWithPrefix(prefix: string | [string, string, string], writer: (s: string) => void, node: Node) {
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: Our emit time keeps creeping upward. I'm wary of making any function more polymorphic than necessary. Can you verify the performance characteristics of this change?

File: src/compiler/types.ts (hunk 3 above)
Comment (rbuckton) [reply src/compiler/types.ts]: Actually, I think the Foo[["bar"]] thing is solved in the dynamicNames PR.

Comment (rbuckton) [reply src/compiler/types.ts]: I wonder if it would have been better to leverage ExpressionWithTypeArguments rather than add this to Identifier.

File: src/compiler/emitter.ts (hunk 4 above)
Comment (weswigham) [reply src/compiler/emitter.ts]: @rbuckton I can refactor all of these callsites to be non-polymorphic (`emitWithPrefix`/`emitWithMultipartPrefix`) pretty easily. I definitely get your concern (hence why there's only one copy of each prefix/postfix array that the multipart prefix/postfix uses). However, I'm _pretty_ certain this will inevitably cause our emit time to increase some _no matter what_, since the node builder inherently has more overhead than direct string building (it actually allocates ASTs, after all) - so it's a matter of identifying everywhere an improvement can be made to keep the change minimal.

File: src/compiler/types.ts (hunk 3 above)
Comment (weswigham) [reply src/compiler/types.ts]: @rbuckton We have this as part of `Identifier` because we reused `EntityName` for the structure which holds this info (which was not allowed to contain arbitrary expressions). With this change, for quickinfo I instead build a real expression tree (instead of an `EntityName`) in order to get the square bracket display for non-identifier names. So I could use `ExpressionWithTypeArguments` instead... but `Identifier` already had the member, and is in use elsewhere for the capture of type arguments for services already. Additionally, the downside of using `ExpressionWithTypeArguments` right now is that it would get auto-parenthesized when it used as the LHS of a `PropertyAccessExpression` (which is both undesirable in this case and would change our output a ton).

Also, there are really only two reasons I didn't fix computed property name to not double-print brackets: 1. To more accurately mimic the old behavior (which we have recorded in baselines) so this is a bit easier to review, and 2. because I'm not 100% sure on how the contents of those square brackets should be parsed (should they be actually parsed to generate an AST as an isolated expression if not a simple number or string literal? Where do we generate symbol names like this, and what constraints are on them?).

File: src/compiler/types.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/types.ts]: I don't have a good sense for how may people rely on our API not changing. I'd prefer to change it, but probably @mhegazy and @DanielRosenwasser have more data on how bad API breaks are. Thoughts?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -2846,7 +2982,14 @@ namespace ts {
             }
 
             function signatureToSignatureDeclarationHelper(signature: Signature, kind: SyntaxKind, context: NodeBuilderContext): SignatureDeclaration {
-                const typeParameters = signature.typeParameters && signature.typeParameters.map(parameter => typeParameterToDeclaration(parameter, context));
+                let typeParameters: TypeParameterDeclaration[];
+                let typeArguments: TypeNode[];
+                if (context.flags & NodeBuilderFlags.WriteTypeArgumentsOfSignature && signature.target && signature.mapper && signature.target.typeParameters) {
+                    typeArguments = signature.target.typeParameters.map(parameter => typeToTypeNodeHelper(instantiateType(parameter, signature.mapper), context));
```
Comment (sandersn) [inline src/compiler/checker.ts]: why not instantiate the whole signature and call this function recursively? Would that call be missing context of some kind?

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2873,15 +3016,17 @@ namespace ts {
...
                     returnTypeNode = createKeywordTypeNode(SyntaxKind.AnyKeyword);
                 }
-                return createSignatureDeclaration(kind, typeParameters, parameters, returnTypeNode);
+                return createSignatureDeclaration(kind, typeParameters, parameters, returnTypeNode, typeArguments);
             }
 
-            function typeParameterToDeclaration(type: TypeParameter, context: NodeBuilderContext): TypeParameterDeclaration {
+            function typeParameterToDeclaration(type: TypeParameter, context: NodeBuilderContext, constraint = getConstraintFromTypeParameter(type)): TypeParameterDeclaration {
+                const savedContextFlags = context.flags;
+                context.flags &= ~NodeBuilderFlags.WriteTypeParametersInQualifiedName; // Avoids potential infinite loop when building for a clodule with a generic
```
Comment (sandersn) [inline src/compiler/checker.ts]: s/clodule/clamespace/? Also, what is the infinite loop case?

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -2898,7 +3043,7 @@ namespace ts {
                         typeToTypeNodeHelper(anyArrayType, context),
                         /*initializer*/ undefined);
                 }
-                const modifiers = parameterDeclaration.modifiers && parameterDeclaration.modifiers.map(getSynthesizedClone);
+                const modifiers = !(context.flags & NodeBuilderFlags.OmitParameterModifiers) && parameterDeclaration.modifiers ? parameterDeclaration.modifiers.map(getSynthesizedClone) : undefined;
```
Comment (sandersn) [inline src/compiler/checker.ts]: this would be easier to read now if you flipped the conditional

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -3018,12 +3129,106 @@ namespace ts {
...
+
+            function lookupTypeParameterNodes(chain: Symbol[], index: number, context: NodeBuilderContext) {
+                Debug.assert(chain && 0 <= index && index < chain.length);
+                const symbol = chain[index];
+                let typeParameterNodes: ReadonlyArray<TypeNode> | ReadonlyArray<TypeParameterDeclaration> | undefined;
+                if (context.flags & NodeBuilderFlags.WriteTypeParametersInQualifiedName && index < (chain.length - 1)) {
+                    const parentSymbol = symbol;
+                    const nextSymbol = chain[index + 1];
+                    if (getCheckFlags(nextSymbol) & CheckFlags.Instantiated) {
+                        typeParameterNodes = mapToTypeNodes(map(getTypeParametersOfClassOrInterface(
```
Comment (sandersn) [inline src/compiler/checker.ts]: help this formatting is unreadable

Comment (sandersn) [reply src/compiler/checker.ts]: I'd recommend extracting the inner call to getTypeParametersOfClassorInterface to a temp, or perhaps just the conditional inside there.

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -3018,12 +3129,106 @@ namespace ts {
                     }
                 }
             }
+
+            function typeParametersToTypeParameterDeclarations(symbol: Symbol, context: NodeBuilderContext) {
+                let typeParameterNodes: ReadonlyArray<TypeParameterDeclaration> | undefined;
+                const targetSymbol = getTargetSymbol(symbol);
+                if (targetSymbol.flags & (SymbolFlags.Class | SymbolFlags.Interface | SymbolFlags.TypeAlias)) {
+                    typeParameterNodes = map(getLocalTypeParametersOfClassOrInterfaceOrTypeAlias(symbol), tp => typeParameterToDeclaration(tp, context));
```
Comment (sandersn) [inline src/compiler/checker.ts]: why is the new map over typeParameterToDeclaration needed here, and why is the map over the symbol's mapper needed in the instantiated case?

Comment (sandersn) [reply src/compiler/checker.ts]: I think it has something to do with the switch to parentSymbol/nextSymbol instead of the weird parentSymbol-only code before.

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -3018,12 +3129,106 @@ namespace ts {
...
+                            (expression as StringLiteral).singleQuote = firstChar === CharacterCodes.singleQuote;
+                        }
+                        if (!expression) {
+                            // TODO: this matches legacy behavior, but results in syntactically invalid identifiers for numeric symbols and computed property name symbols
+                            expression = setEmitFlags(createIdentifier(symbolName, typeParameterNodes), EmitFlags.NoAsciiEscaping);
+                            expression.symbol = symbol;
+                        }
+                        return createElementAccess(createExpressionFromSymbolChain(chain, index - 1), expression);
+                    }
+                    else {
```
Comment (sandersn) [inline src/compiler/checker.ts]: I would flip this to put the common case first, something like `if (isLegalPropertyAccess || index === 0) ...`

File: src/compiler/emitter.ts [hunk 11]
```diff
@@ -476,7 +481,9 @@ namespace ts {
 
         function emitMappedTypeParameter(node: TypeParameterDeclaration): void {
             emit(node.name);
-            write(" in ");
+            writeSpace(" ");
```
Comment (sandersn) [inline src/compiler/emitter.ts]: why the argument? what other kinds of spaces are there?

File: src/services/signatureHelp.ts [hunk 12]
```diff
@@ -375,23 +375,59 @@ namespace ts.SignatureHelp {
...
             else {
                 isVariadic = candidateSignature.hasRestParameter;
-                const typeParameterParts = mapToDisplayParts(writer =>
-                    typeChecker.getSymbolDisplayBuilder().buildDisplayForTypeParametersAndDelimiters(candidateSignature.typeParameters, writer, invocation));
+                const typeParameterParts = mapToDisplayParts(writer => {
+                    if (candidateSignature.typeParameters && candidateSignature.typeParameters.length) {
+                        writer.writePunctuation("<");
+                        let hadPrevious = false;
+                        for (const param of candidateSignature.typeParameters) {
+                            if (hadPrevious) {
```
Comment (sandersn) [inline src/services/signatureHelp.ts]: it's a tiny bit concerning that the dissolution of the symbol display builder caused the dispersal of very specific display code over the code base. Specifically, I saw three new copies of this parameter-display loop. Is there any way to consolidate this somewhere else?

File: src/compiler/emitter.ts (hunk 4 above)
Comment (sandersn) [reply src/compiler/emitter.ts]: I'd go so far as to say that, since there are only 3 triples right now, I'd create dedicated functions for the first two triples (used 10 and 6 times, respectively, something like `emitAsTypeAnnotation`) and inline the last, which is only used once.

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -2512,7 +2628,7 @@ namespace ts {
                     // Ignore constraint/default when creating a usage (as opposed to declaration) of a type parameter.
                     return createTypeReferenceNode(name, /*typeArguments*/ undefined);
                 }
-                if (!inTypeAlias && type.aliasSymbol && isTypeSymbolAccessible(type.aliasSymbol, context.enclosingDeclaration)) {
+                if (!inTypeAlias && type.aliasSymbol && (context.flags & NodeBuilderFlags.UseAliasDefinedOutsideCurrentScope || isTypeSymbolAccessible(type.aliasSymbol, context.enclosingDeclaration))) {
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/checker.ts]: Yep - the symbol display builder supported these options for the quick info or signature help display, so the node builder needed to do so as well in order to support the same output.

File: src/compiler/checker.ts (hunk 5 above)
Comment (weswigham) [reply src/compiler/checker.ts]: When we instantiate a signature, we instantiate the parameter types and the return type - type parameters get copied to new copies of themselves, but are not instantiated with the mapper until needed (hence why it needs to be handled on the consuming end). If we did instantiate them, they would no longer be guaranteed to be `TypeParameter`s, and would no longer be able to be stored in the signature's `typeParameter` array. 😉

File: src/compiler/checker.ts (hunk 6 above)
Comment (weswigham) [reply src/compiler/checker.ts]: Printing quickinfo for `Bar` where `Bar` is a class in `namespace Foo` where `Foo` is merged with class `Foo<T extends Foo.Bar>`.

File: src/compiler/checker.ts (hunk 9 above)
Comment (weswigham) [reply src/compiler/checker.ts]: So before `parentSymbol` was the symbol left of the current symbol in the chain - now `parentSymbol` is the current symbol, and `nextSymbol` is the one to the right of the current one. We say "if the symbol on the right is instantiated, then it must have come from an instantiated version of the symbol on the left - get its mapper and apply it to the type parameters of the symbol on the left". It was the same in the symbol display builder, except it was more straightforwardly written, since it built the output string from left to right, whereas we build the node tree from right to left.

File: src/compiler/emitter.ts (hunk 11 above)
Comment (weswigham) [reply src/compiler/emitter.ts]: The distinction is for the symbol writer-style-usage. AFAIK, you're right and `" "` is the only whitespace we ever call it with. I'll push the literal to the edge where we interact with the writer.

File: src/services/signatureHelp.ts (hunk 12 above)
Comment (weswigham) [reply src/services/signatureHelp.ts]: Out emitter cannot emit a list directly, as a list is not a node kind... so these exist mostly to work around how I can't pass a `NodeArray` and list format into the emitter directly... Would it make sense to expose a `writeList`/`printList` alongside `writeNode`/`printNode`? Thinking on it, it probably does.

**Comment by DanielRosenwasser** [comment]
@LucasSloan @evmar @alexeagle https://github.com/angular/tsickle/blob/423081590c008e239868c4c657299630793f16b6/src/type-translator.ts#L217 seems to use this API.

Given the problems you ran into [described here](https://github.com/angular/tsickle/blob/423081590c008e239868c4c657299630793f16b6/src/type-translator.ts#L163-L164) we believe that `symbolToString` should be sufficiently configurable so that you can easily migrate off of `getSymbolDisplayBuilder`.

**Comment by weswigham** [comment]
Or rather `symbolToString` should be configurable enough if I have clearance to replace `SymbolFormatFlags` with `NodeBuilderFlags` (as the fully-qualified bit is a node builder or type format flag and not a symbol format flag) - failing that, `createPrinter` combined with and the new `symbolToEntityName` or `symbolToExpression` allows the same configurable `symbolToString` behavior.

**Comment by weswigham** [comment]
@DanielRosenwasser This now includes a shim for `getSymbolDisplayBuilder` which should be compatible with the old one's behavior (but implemented with nodeBuilder APIs), which can be used while the function is deprecated.

File: src/compiler/types.ts (hunk 3 above)
Comment (aozgaa) [inline src/compiler/types.ts]: Should this comment now say it is used for emitting diagnostics and quickinfo? Are there any other cases where this might be used? Do we have some way of asserting this isn't misused in the emitter (eg: assert it's undefined when emitting a declaration file)?

File: src/compiler/types.ts [hunk 14]
```diff
@@ -693,6 +693,8 @@ namespace ts {
         typeParameters?: NodeArray<TypeParameterDeclaration>;
         parameters: NodeArray<ParameterDeclaration>;
         type: TypeNode | undefined;
+        /* @internal */
+        typeArguments?: NodeArray<TypeNode>; // Used for quick info, replaces typeParameters for instantiated signatures
```
Comment (aozgaa) [inline src/compiler/types.ts]: nit: put `@internal` annotation on the same line.

File: src/compiler/types.ts [hunk 15]
```diff
@@ -4636,4 +4689,86 @@ namespace ts {
...
+        MultiLine = 1 << 0,             // Prints the list on multiple lines.
+        PreserveLines = 1 << 1,         // Prints the list using line preservation if possible.
+        LinesMask = SingleLine | MultiLine | PreserveLines,
+
+        // Delimiters
+        NotDelimited = 0,               // There is no delimiter between list items (default).
+        BarDelimited = 1 << 2,          // Each list item is space-and-bar (" |") delimited.
+        AmpersandDelimited = 1 << 3,    // Each list item is space-and-ampersand (" &") delimited.
+        CommaDelimited = 1 << 4,        // Each list item is comma (",") delimited.
+        DelimitersMask = BarDelimited | AmpersandDelimited | CommaDelimited,
```
Comment (aozgaa) [inline src/compiler/types.ts]: Should `NotDelimited` be part of the mask?

File: src/compiler/types.ts [hunk 16]
```diff
@@ -4636,4 +4689,86 @@ namespace ts {
...
+
+        NoSpaceIfEmpty = 1 << 18,       // If the literal is empty, do not add spaces between braces.
+        SingleElement = 1 << 19,
+
+        // Precomputed Formats
+        Modifiers = SingleLine | SpaceBetweenSiblings | NoInterveningComments,
+        HeritageClauses = SingleLine | SpaceBetweenSiblings,
+        SingleLineTypeLiteralMembers = SingleLine | SpaceBetweenBraces | SpaceBetweenSiblings | Indented,
+        MultiLineTypeLiteralMembers = MultiLine | Indented,
+
```
Comment (aozgaa) [inline src/compiler/types.ts]: nit: why the extra newline?

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 17]
```diff
@@ -1762,30 +1773,42 @@ declare namespace ts {
...
+        /** @deprecated */ buildTypeParameterDisplay(tp: TypeParameter, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
+        /** @deprecated */ buildTypePredicateDisplay(predicate: TypePredicate, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
+        /** @deprecated */ buildTypeParameterDisplayFromSymbol(symbol: Symbol, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
+        /** @deprecated */ buildDisplayForParametersAndDelimiters(thisParameter: Symbol, parameters: Symbol[], writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
+        /** @deprecated */ buildDisplayForTypeParametersAndDelimiters(typeParameters: TypeParameter[], writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
+        /** @deprecated */ buildReturnTypeDisplay(signature: Signature, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags): void;
     }
+    /**
+     * @deprecated Migrate to EmitTextWriter
```
Comment (aozgaa) [inline tests/baselines/reference/api/tsserverlibrary.d.ts]: Why deprecate this?

File: src/compiler/emitter.ts [hunk 18]
```diff
@@ -2349,7 +2484,14 @@ namespace ts {
             emitList(parentNode, typeArguments, ListFormat.TypeArguments);
         }
 
-        function emitTypeParameters(parentNode: Node, typeParameters: NodeArray<TypeParameterDeclaration>) {
+        function emitTypeParameters(parentNode: SignatureDeclaration | InterfaceDeclaration | TypeAliasDeclaration | ClassDeclaration | ClassExpression, typeParameters: NodeArray<TypeParameterDeclaration>) {
+            if (parentNode.kind !== SyntaxKind.InterfaceDeclaration &&
```
Comment (aozgaa) [inline src/compiler/emitter.ts]: Should we throw an exception if the type arguments are defined for these syntax kinds?

File: src/compiler/emitter.ts [hunk 19]
```diff
@@ -2302,35 +2420,52 @@ namespace ts {
...
         }
 
-        function emitExpressionWithPrefix(prefix: string, node: Node) {
-            emitNodeWithPrefix(prefix, node, emitExpression);
+        function emitExpressionWithPrefix(prefix: string, writer: (s: string) => void, node: Node) {
+            emitNodeWithPrefix(prefix, writer, node, emitExpression);
         }
 
-        function emitNodeWithPrefix(prefix: string, node: Node, emit: (node: Node) => void) {
+        function emitNodeWithPrefix(prefix: string, writer: (s: string) => void, node: Node, emit: (node: Node) => void) {
```
Comment (aozgaa) [inline src/compiler/emitter.ts]: Could we add a default initializer for `emit` to eliminate the similarly-named `emitWithPrefix`?

Comment (aozgaa) [inline src/compiler/emitter.ts]: Would it make sense to name the second parameter `prefixWriter` to emphasize this writer is used only for `prefix`?

**Comment by weswigham** [comment]
Ping @DanielRosenwasser and @sandersn for a rereview. Best I can tell, this shouldn't break too many people anymore. The only break left is that `SymbolWriter` is gone - its contents merged into `EmitTextWriter` (can't really keep it around as an unused base type as then implementers would need two close/reset methods). A [cursory search](https://github.com/search?utf8=%E2%9C%93&q=SymbolWriter+language%3ATypeScript+-filename%3Atypescript.d.ts+-filename%3Atypes.ts+-filename%3Aindex.d.ts+-filename%3Ats.d.ts+-filename%3Atype-translator.ts+-filename%3Autilities.ts+-filename%3AdeclarationEmitter.ts+-filename%3Aservices.ts&type=Code) that removes our own sources (and tcikle) reveals that it's unlikely to affect any active projects.

File: src/compiler/emitter.ts (hunk 18 above)
Comment (weswigham) [reply src/compiler/emitter.ts]: They won't be because the member doesn't exist on them. Typechecker's got our back here; the reason the if statement is here is to safely narrow to `SignatureDeclaration`, which are the only types which could have them member defined on them.

File: src/compiler/emitter.ts (hunk 19 above)
Comment (weswigham) [reply src/compiler/emitter.ts]: Sure.

File: src/compiler/types.ts (hunk 16 above)
Comment (weswigham) [reply src/compiler/types.ts]: It was there in the version of this which used to be within `emitter.ts`. I assume to signify some logical separation between the literal member formats and the other formats.

File: src/compiler/types.ts (hunk 15 above)
Comment (weswigham) [reply src/compiler/types.ts]: It's `0`, so..... it wouldn't actually be part of the mask.

File: src/compiler/types.ts (hunk 14 above)
Comment (weswigham) [reply src/compiler/types.ts]: done

File: src/compiler/types.ts (hunk 3 above)
Comment (weswigham) [reply src/compiler/types.ts]: Re: the comment. Sure, done. Re: asserting misuse: It is only set when the caller to the node builder sets an explicit "I want type parameters on Identifiers" flag. It's not an easy thing to get by accident.

File: src/compiler/emitter.ts (hunk 19 above)
Comment (weswigham) [reply src/compiler/emitter.ts]: Sure. `emitWithPrefix` is only used in 3 places anyway.

**Comment by weswigham** [comment]
@DanielRosenwasser @mhegazy I've merged with master and tightened up the API changes - now `EmitTextWriter` is still `@internal`, and the only removal happens to be `DisplayPartsSymbolWriter` - which, while the interface was exposed, was impossible to construct using our API, and was not required for consumption anywhere in our API, so should have been `@internal` all along.

**Comment by mhegazy** [comment]
@rbuckton any comments?

File: src/compiler/emitter.ts [hunk 20]
```diff
@@ -2591,57 +2737,117 @@ namespace ts {
...
         }
 
+        function writeSymbol(s: string, sym: Symbol) {
+            commitPendingSemicolon();
+            writer.writeSymbol(s, sym);
+        }
+
+        function writePunctuation(s: string) {
+            commitPendingSemicolon();
+            if (printerOptions.omitTrailingSemicolon && s === ";") {
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: Add a `writeSemicolon()` function for this case (and replace every `writePunctuation(";")`) so we don't have to observe the cost of the lookup and string comparison for every punctuation mark.

File: src/compiler/types.ts (hunk 16 above)
Comment (rbuckton) [reply src/compiler/types.ts]: I don't think it serves a specific purpose.

File: src/compiler/visitor.ts [hunk 21]
```diff
@@ -222,7 +222,7 @@ namespace ts {
             // Names
 
             case SyntaxKind.Identifier:
-                return updateIdentifier(<Identifier>node, nodesVisitor((<Identifier>node).typeArguments, visitor, isTypeNode));
+                return updateIdentifier(<Identifier>node, nodesVisitor((<Identifier>node).typeArguments, visitor, or(isTypeNode, isTypeParameterDeclaration)));
```
Comment (rbuckton) [inline src/compiler/visitor.ts]: I understand your reasoning for adding `or`, but I still believe `or(isTypeNode, ...)` should be pulled out into a separate function.

File: src/services/signatureHelp.ts [hunk 22]
```diff
@@ -375,23 +375,41 @@ namespace ts.SignatureHelp {
                 const typeParameters = (candidateSignature.target || candidateSignature).typeParameters;
                 signatureHelpParameters = typeParameters && typeParameters.length > 0 ? map(typeParameters, createSignatureHelpParameterForTypeParameter) : emptyArray;
                 suffixDisplayParts.push(punctuationPart(SyntaxKind.GreaterThanToken));
-                const parameterParts = mapToDisplayParts(writer =>
-                    typeChecker.getSymbolDisplayBuilder().buildDisplayForParametersAndDelimiters(candidateSignature.thisParameter, candidateSignature.parameters, writer, invocation));
+                const parameterParts = mapToDisplayParts(writer => {
+                    const printer = createPrinter({ removeComments: true });
```
Comment (rbuckton) [inline src/services/signatureHelp.ts]: `Printer` resets after each use. Rather than reallocating a `printer` each time, I'd allocate `printer` once and reuse it.

File: src/services/signatureHelp.ts [hunk 23]
```diff
@@ -375,23 +375,41 @@ namespace ts.SignatureHelp {
...
+                });
                 addRange(suffixDisplayParts, parameterParts);
             }
             else {
                 isVariadic = candidateSignature.hasRestParameter;
-                const typeParameterParts = mapToDisplayParts(writer =>
-                    typeChecker.getSymbolDisplayBuilder().buildDisplayForTypeParametersAndDelimiters(candidateSignature.typeParameters, writer, invocation));
+                const typeParameterParts = mapToDisplayParts(writer => {
+                    if (candidateSignature.typeParameters && candidateSignature.typeParameters.length) {
+                        const printer = createPrinter({ removeComments: true });
```
Comment (rbuckton) [inline src/services/signatureHelp.ts]: As above, you only need one `printer`.

File: src/services/signatureHelp.ts [hunk 24]
```diff
@@ -427,8 +448,11 @@ namespace ts.SignatureHelp {
         }
 
         function createSignatureHelpParameterForTypeParameter(typeParameter: TypeParameter): SignatureHelpParameter {
-            const displayParts = mapToDisplayParts(writer =>
-                typeChecker.getSymbolDisplayBuilder().buildTypeParameterDisplay(typeParameter, writer, invocation));
+            const displayParts = mapToDisplayParts(writer => {
+                const printer = createPrinter({ removeComments: true });
```
Comment (rbuckton) [inline src/services/signatureHelp.ts]: Again, `Printer` is designed to be reused when passed the same set of initial arguments. No sense in recreating it all the time.

File: src/services/utilities.ts [hunk 25]
```diff
@@ -1123,8 +1124,18 @@ namespace ts {
...
             writeSymbol,
             writeLine,
+            write: unknownWrite,
+            writeTextOfNode: unknownWrite,
+            getText: () => "",
+            getTextPos: () => 0,
+            getColumn: () => 0,
+            getLine: () => 0,
+            isAtStartOfLine: () => false,
+            rawWrite: noop,
```
Comment (rbuckton) [inline src/services/utilities.ts]: Should this be `noop` or should it raise an error when called?

File: src/compiler/types.ts [hunk 26]
```diff
@@ -2651,8 +2662,13 @@ namespace ts {
         getTypeAtLocation(node: Node): Type;
         getTypeFromTypeNode(node: TypeNode): Type;
         signatureToString(signature: Signature, enclosingDeclaration?: Node, flags?: TypeFormatFlags, kind?: SignatureKind): string;
+        /* @internal */ signatureToString(signature: Signature, enclosingDeclaration?: Node, flags?: TypeFormatFlags, kind?: SignatureKind, writer?: EmitTextWriter): string;
```
Comment (rbuckton) [inline src/compiler/types.ts]: Rather than overloading these, why not have:

```ts
  /* @internal */ writeSignature(writer: EmitTextWriter, signature: Signature, ...): void;
  /* @internal */ writeType(writer: EmitTextWriter, type: Type, ...): void;
  /* @internal */ writeSymbol(writer: EmitTextWriter, symbol: Symbol, ...): void;
  /* @internal */ writeTypePredicate(writer: EmitTextWriter, predicate: TypePredicate, ...): void;
```

That is similar to how `Printer` splits up internal and public members.

File: src/compiler/types.ts [hunk 27]
```diff
@@ -2841,6 +2864,9 @@ namespace ts {
         // eg.  module m { export class c { } } import x = m.c;
         // When this flag is specified m.c will be used to refer to the class instead of alias symbol x
         UseOnlyExternalAliasing = 0x00000002,
+
+        // Build symbol name using property accesses and element accesses, instead of an entity name
+        AllowAnyNodeKind = 0x00000004,
```
Comment (rbuckton) [inline src/compiler/types.ts]: The name `AllowAnyNodeKind` doesn't seem to align with the intended use described in the comment. Perhaps a better description or a more descriptive name would be better here.

File: src/services/utilities.ts (hunk 25 above)
Comment (weswigham) [reply src/services/utilities.ts]: I've changed it to raise an error, since `rawWrite` is only used to print comments, and the display part writer should only be used when comment printing is disabled.

File: src/compiler/types.ts (hunk 27 above)
Comment (weswigham) [reply src/compiler/types.ts]: Updated the description

File: src/compiler/emitter.ts [hunk 28]
```diff
@@ -286,16 +286,21 @@ namespace ts {
         let tempFlags: TempFlags; // TempFlags for the current name generation scope.
         let writer: EmitTextWriter;
         let ownWriter: EmitTextWriter;
+        let write = writeBase;
+        let pendingSemicolon = false;
+        const syntheticParent: TextRange = { pos: -1, end: -1 };
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/emitter.ts]: I've moved `pos` and `end` before the other fields - they were already first in the services constructor implementation for `Node`, so I think this only changes `tsc`.

**Comment by weswigham** [comment]
@rbuckton I've finally gotten around to syncing this and implementing your comments. Is there anything else you'd like to see?

**Comment by mhegazy** [comment]
@weswigham can you please refresh, sync with @rbuckton and merge
